File: android/src/main/java/org/ligi/gobandroid_hd/ui/sgf_listing/GoLink.java
package org.ligi.gobandroid_hd.ui.sgf_listing;

import org.ligi.axt.AXT;
import org.ligi.gobandroid_hd.FileEncodeDetector;
import org.ligi.gobandroid_hd.logic.GoGame;
import org.ligi.tracedroid.logging.Log;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;

public class GoLink {

    public static boolean isGoLink(String fname) {
        return fname.endsWith(".golink");
    }

    public static boolean isGoLink(File file) {
        return file.getName().endsWith(".golink");
    }


    private String fname = "";
    private int move_pos = 0;

    public GoLink(String fname) {
        this(new File(fname.replace("file://", "")));
    }

    public GoLink(File file) {

        try {
            String go_lnk = AXT.at(file).readToString();
            go_lnk = go_lnk.replace("\n", "").replace("\r", "");
            fname = go_lnk; // backup
            String[] arr_content = go_lnk.split(":#");
            fname = arr_content[0];
            fname = fname.replace("file://", "");
            move_pos = Integer.parseInt(arr_content[1]);
        } catch (Exception e) {
        }
    }

    /**
     * returns the move Depth/pos
     * should only be used for displaying
     *
     * @return
     */
    public int getMoveDepth() {
        return move_pos;
    }


    public boolean linksToDirectory() {
        return new File(fname).isDirectory();
    }

    /**
     * TODO care for remote content
     *
     * @return
     */
    public String getSGFString() {
        try {
            return AXT.at(new File(fname)).readToString(FileEncodeDetector.detect(fname));
        } catch (IOException e) {
            return "";
        }
    }

    public String getFileName() {
        return fname;
    }

    public static void saveGameToGoLink(GoGame game, String golink_path, String golink_fname) {

        int move_pos = game.getActMove().getMovePos();

        File f = new File(golink_path);

        if (!f.isDirectory())
            f.mkdirs();

        try {
            f.createNewFile();
            f = new File(golink_path + "/" + golink_fname);

            FileWriter sgf_writer = new FileWriter(f);

            BufferedWriter out = new BufferedWriter(sgf_writer);

            out.write(game.getMetaData().getFileName() + ":#" + move_pos);
            out.close();
            sgf_writer.close();
        } catch (IOException e) {
            Log.i("" + e);
        }
    }
}

File: android/src/main/java/org/ligi/gobandroid_hd/ui/sgf_listing/GoLink.kt
package org.ligi.gobandroid_hd.ui.sgf_listing

import org.ligi.gobandroid_hd.logic.GoGame
import timber.log.Timber
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException

class GoLink(file: File) {


    var fileName = ""
        private set
    /**
     * returns the move Depth/pos
     * should only be used for displaying

     * @return
     */
    var moveDepth = 0
        private set

    constructor(fname: String) : this(File(fname.replace("file://", ""))) {
    }

    init {

        try {
            var go_lnk = file.bufferedReader().readText()
            go_lnk = go_lnk.replace("\n", "").replace("\r", "")
            fileName = go_lnk // backup
            val arr_content = go_lnk.split(":#".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
            fileName = arr_content[0]
            fileName = fileName.replace("file://", "")
            moveDepth = Integer.parseInt(arr_content[1])
        } catch (e: Exception) {
        }

    }


    fun linksToDirectory(): Boolean {
        return File(fileName).isDirectory
    }

    /**
     * TODO care for remote content

     * @return
     */
    val sgfString: String
        get() {
            try {
                return File(fileName).bufferedReader().readText()
            } catch (e: IOException) {
                return ""
            }

        }

    companion object {

        fun isGoLink(fname: String): Boolean {
            return fname.endsWith(".golink")
        }

        fun isGoLink(file: File): Boolean {
            return file.name.endsWith(".golink")
        }

        fun saveGameToGoLink(game: GoGame, golink_path: File, golink_fname: String) {
            val move_pos = game.actMove.movePos

            if (!golink_path.isDirectory)
                golink_path.mkdirs()

            try {
                val golink_file = File(golink_path.toString() + "/" + golink_fname)

                val sgf_writer = FileWriter(golink_file)

                val out = BufferedWriter(sgf_writer)

                out.write(game.metaData.fileName + ":#" + move_pos)
                out.close()
                sgf_writer.close()
            } catch (e: IOException) {
                Timber.i("" + e)
            }

        }
    }
}
